programs/host/frontend.c

```diff
@@ -0,0 +1,943 @@
+ * parameters and reply MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED to the next
+ * function then set the status back to OK.
+ * In case of any I/O error, set ctx->status to
+ * MBEDTLS_SERIALIZE_STATUS_DEAD. */
+static int mbedtls_serialize_pull( mbedtls_serialize_context_t *ctx )
+{
+    int ret;
+    uint8_t header[4];
+    if( ctx->status == MBEDTLS_SERIALIZE_STATUS_DEAD )
+    {
+        DBG( "already dead" );
+        return( MBEDTLS_ERR_SERIALIZE_RECEIVE );
+    }
+    if( ( ret = mbedtls_serialize_read( ctx, header, sizeof( header ) ) ) != 0 )
+    {
+        DBG( "receive failure -> dead" );
+        ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+        return( ret );
+    }
+    switch( header[0] )
+    {
+    case MBEDTLS_SERIALIZE_TYPE_PUSH:
+    {
+        size_t length = header[1] << 16 | header[2] << 8 | header[3];
+        mbedtls_serialize_item_t *item;
+        DBG( "received push length=%zu", length );
+        if( ctx->status == MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY )
+        {
+            /* While we're out of memory, keep reading arguments but discard
+               them. */
+            while( length > sizeof( header ) )
+            {
+                if( ( ret = mbedtls_serialize_read( ctx, header, sizeof( header ) ) ) != 0 )
+                {
+                    DBG( "failed to read input with %zu bytes remaining -> dead", length );
+                    ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+                    return( ret );
+                }
+                length -= sizeof( header );
+            }
+            if( ( ret = mbedtls_serialize_read( ctx, header, length ) ) != 0 )
+            {
+                DBG( "failed to read input with %zu bytes remaining -> dead", length );
+                ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+                return( ret );
+            }
+            return( MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED );
+        }
+        item = alloc_item( length );
+        if( item == NULL )
+        {
+            DBG( "failed to allocate %zu bytes for input", length );
+            ctx->status = MBEDTLS_SERIALIZE_STATUS_OUT_OF_MEMORY;
+            return( MBEDTLS_ERR_SERIALIZE_ALLOC_FAILED );
+        }
+        if( ( ret = mbedtls_serialize_read( ctx, item_buffer( item ), length ) ) != 0 )
+        {
+            DBG( "failed to read %zu-byte input -> dead", length );
+            ctx->status = MBEDTLS_SERIALIZE_STATUS_DEAD;
+            return( ret );
+        }
+        DBG( "successfully read %zu-byte input", length );
+        item->next = ctx->stack;
+        ctx->stack = item;
+        return( 0 );
+    }
+
+    case MBEDTLS_SERIALIZE_TYPE_EXECUTE:
+    {
+        uint32_t function = header[1] << 16 | header[2] << 8 | header[3];
+        uint32_t status;
+        mbedtls_serialize_uint32_t status_item = {{NULL, 4}, {0}};
+        uint8_t *status_data = item_buffer( &status_item.meta );
+        mbedtls_serialize_item_t *outputs[1 + 16] = {&status_item.meta};
```

Comment:
This only initializes the first item of `outputs`, while the others are unspecified. The documentation of `mbedtls_serialize_perform` and the `outputs[i] != NULL` checks below indicate that `outputs` should be `NULL` initialized.